File: app/src/main/java/sk/styk/martin/apkanalyzer/model/statistics/MathStatistics.java
package sk.styk.martin.apkanalyzer.model.statistics;

import android.os.Parcel;
import android.os.Parcelable;

import java.math.BigDecimal;

/**
 * Math Statistics collected about attributes of Apk file
 * <p>
 * @author Martin Styk
 * @version 28.07.2017.
 */
public class MathStatistics implements Parcelable {

    public static final Parcelable.Creator<MathStatistics> CREATOR = new Parcelable.Creator<MathStatistics>() {
        @Override
        public MathStatistics createFromParcel(Parcel source) {
            return new MathStatistics(source);
        }

        @Override
        public MathStatistics[] newArray(int size) {
            return new MathStatistics[size];
        }
    };
    private BigDecimal arithmeticMean;
    private BigDecimal median;
    private BigDecimal max;
    private BigDecimal min;
    private BigDecimal variance;
    private BigDecimal deviation;


    protected MathStatistics(BigDecimal arithmeticMean,
                           BigDecimal median,
                           BigDecimal max,
                           BigDecimal min,
                           BigDecimal variance,
                           BigDecimal deviation) {
        this.arithmeticMean = arithmeticMean;
        this.median = median;
        this.max = max;
        this.min = min;
        this.variance = variance;
        this.deviation = deviation;
    }

    protected MathStatistics(Parcel in) {
        this.arithmeticMean = (BigDecimal) in.readSerializable();
        this.median = (BigDecimal) in.readValue(Long.class.getClassLoader());
        this.max = (BigDecimal) in.readSerializable();
        this.min = (BigDecimal) in.readSerializable();
        this.variance = (BigDecimal) in.readSerializable();
        this.deviation = (BigDecimal) in.readSerializable();
    }


    public BigDecimal getArithmeticMean() {
        return arithmeticMean;
    }

    public BigDecimal getMedian() {
        return median;
    }

    public BigDecimal getMax() {
        return max;
    }

    public BigDecimal getMin() {
        return min;
    }

    public BigDecimal getVariance() {
        return variance;
    }

    public BigDecimal getDeviation() {
        return deviation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MathStatistics that = (MathStatistics) o;

        if (arithmeticMean != null ? !arithmeticMean.equals(that.arithmeticMean) : that.arithmeticMean != null)
            return false;
        if (median != null ? !median.equals(that.median) : that.median != null) return false;
        if (max != null ? !max.equals(that.max) : that.max != null) return false;
        if (min != null ? !min.equals(that.min) : that.min != null) return false;
        if (variance != null ? !variance.equals(that.variance) : that.variance != null)
            return false;
        return deviation != null ? deviation.equals(that.deviation) : that.deviation == null;

    }

    @Override
    public int hashCode() {
        int result = arithmeticMean != null ? arithmeticMean.hashCode() : 0;
        result = 31 * result + (median != null ? median.hashCode() : 0);
        result = 31 * result + (max != null ? max.hashCode() : 0);
        result = 31 * result + (min != null ? min.hashCode() : 0);
        result = 31 * result + (variance != null ? variance.hashCode() : 0);
        result = 31 * result + (deviation != null ? deviation.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "MathStatistics{" +
                "arithmeticMean=" + arithmeticMean +
                ", median=" + median +
                ", max=" + max +
                ", min=" + min +
                ", variance=" + variance +
                ", deviation=" + deviation +
                '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeSerializable(this.arithmeticMean);
        dest.writeValue(this.median);
        dest.writeSerializable(this.max);
        dest.writeSerializable(this.min);
        dest.writeSerializable(this.variance);
        dest.writeSerializable(this.deviation);
    }
}

File: core/app-statistics/src/main/kotlin/sk/styk/martin/apkanalyzer/core/appstatistics/model/MathStatistics.kt
package sk.styk.martin.apkanalyzer.core.appstatistics.model

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.math.BigDecimal
import java.util.Arrays
import kotlin.math.sqrt

@Parcelize
data class MathStatistics(
    val arithmeticMean: BigDecimal,
    var median: BigDecimal,
    var max: BigDecimal,
    var min: BigDecimal,
    var variance: BigDecimal,
    var deviation: BigDecimal,
) : Parcelable


internal fun FloatArray.toMathStats(): MathStatistics {
    if (this.isEmpty()) {
        throw IllegalArgumentException()
    }

    Arrays.sort(this)

    fun computeMean(): Double = sum() / size.toDouble()

    fun computeMedian(): Float {
        return if (size % 2 == 0) {
            (this[size / 2 - 1] + this[size / 2]) / 2.0f
        } else {
            this[size / 2]
        }
    }

    fun computeVariance(mean: Float): Double {
        var temp = 0.0
        for (a in this)
            temp += (a - mean) * (a - mean)
        return temp / (size - 1)
    }


    val arithmeticMean = BigDecimal(computeMean())
    val variance = BigDecimal(computeVariance(arithmeticMean.toFloat()))
    val deviation = BigDecimal(sqrt(variance.toFloat().toDouble()))

    return MathStatistics(
        arithmeticMean = arithmeticMean,
        median = BigDecimal(computeMedian().toDouble()),
        max = BigDecimal(this[this.size - 1].toDouble()),
        min = BigDecimal(this[0].toDouble()),
        variance = variance,
        deviation = deviation,
    )
}